File: src/slagalica/gui/SlagalicaGUI.java
package slagalica.gui;

import java.awt.EventQueue;
import java.awt.Font;
import java.awt.GridBagConstraints;
import java.awt.GridBagLayout;
import java.awt.Insets;
import java.awt.event.ActionListener;
import java.io.FileNotFoundException;
import java.io.IOException;

import javax.imageio.ImageIO;
import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JList;
import javax.swing.JScrollPane;
import javax.swing.JTextField;

import slagalica.core.Dictionary;

public class SlagalicaGUI {

	private static final String	ICON_PATH		= "images/icon.png";
	private static final String	DICTIONARY_PATH	= "data/sr-cyrl.wl";

	private JFrame				frmSlagalica;
	private JTextField			textField;

	private Dictionary			serbianDictionary;
	private WordListModel		wordListModel;

	/**
	 * Launch the application.
	 */
	public static void main(String[] args) {
		EventQueue.invokeLater(new Runnable() {
			public void run() {
				try {
					SlagalicaGUI window = new SlagalicaGUI();
					window.frmSlagalica.setVisible(true);
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
		});
	}

	/**
	 * Create the application.
	 * 
	 * @throws FileNotFoundException
	 */
	public SlagalicaGUI() throws FileNotFoundException {
		initialize();

		serbianDictionary = new Dictionary(DICTIONARY_PATH);
		wordListModel.setValues(serbianDictionary.findAllWords(null));
	}

	/**
	 * Initialize the contents of the frame.
	 */
	private void initialize() {
		frmSlagalica = new JFrame();
		frmSlagalica.setTitle("Slagalica");
		frmSlagalica.getContentPane().setFont(new Font("Segoe UI", Font.PLAIN, 12));
		frmSlagalica.setBounds(100, 100, 578, 405);
		frmSlagalica.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		ActionListener findWords = new FindWordActionListener(this);
		GridBagLayout gridBagLayout = new GridBagLayout();
		gridBagLayout.columnWidths = new int[] { 100, 300, 100, 0 };
		gridBagLayout.rowHeights = new int[] { 0, 0, 0 };
		gridBagLayout.columnWeights = new double[] { 0.0, 2.0, 0.0, Double.MIN_VALUE };
		gridBagLayout.rowWeights = new double[] { 0.0, 1.0, Double.MIN_VALUE };
		frmSlagalica.getContentPane().setLayout(gridBagLayout);

		JLabel lblUnesiteSlova = new JLabel("Унесите слова:");
		lblUnesiteSlova.setFont(new Font("Segoe UI", Font.PLAIN, 12));
		GridBagConstraints gbc_lblUnesiteSlovaZa = new GridBagConstraints();
		gbc_lblUnesiteSlovaZa.insets = new Insets(0, 5, 5, 5);
		gbc_lblUnesiteSlovaZa.anchor = GridBagConstraints.WEST;
		gbc_lblUnesiteSlovaZa.gridx = 0;
		gbc_lblUnesiteSlovaZa.gridy = 0;
		frmSlagalica.getContentPane().add(lblUnesiteSlova, gbc_lblUnesiteSlovaZa);

		textField = new JTextField();
		textField.addActionListener(findWords);
		textField.setFont(new Font("Segoe UI", Font.PLAIN, 12));
		GridBagConstraints gbc_textField = new GridBagConstraints();
		gbc_textField.insets = new Insets(2, 0, 5, 5);
		gbc_textField.fill = GridBagConstraints.BOTH;
		gbc_textField.gridx = 1;
		gbc_textField.gridy = 0;
		frmSlagalica.getContentPane().add(textField, gbc_textField);
		textField.setColumns(10);

		JButton btnNewButton = new JButton("Пронађи речи");
		btnNewButton.addActionListener(findWords);
		btnNewButton.setFont(new Font("Segoe UI", Font.PLAIN, 12));
		GridBagConstraints gbc_btnNewButton = new GridBagConstraints();
		gbc_btnNewButton.insets = new Insets(0, 5, 5, 0);
		gbc_btnNewButton.fill = GridBagConstraints.BOTH;
		gbc_btnNewButton.gridx = 2;
		gbc_btnNewButton.gridy = 0;
		frmSlagalica.getContentPane().add(btnNewButton, gbc_btnNewButton);

		JScrollPane scrollPane = new JScrollPane();
		scrollPane.setViewportBorder(null);
		GridBagConstraints gbc_scrollPane = new GridBagConstraints();
		gbc_scrollPane.fill = GridBagConstraints.BOTH;
		gbc_scrollPane.gridwidth = 3;
		gbc_scrollPane.insets = new Insets(0, 5, 5, 5);
		gbc_scrollPane.gridx = 0;
		gbc_scrollPane.gridy = 1;
		frmSlagalica.getContentPane().add(scrollPane, gbc_scrollPane);

		JList<String> list = new JList<String>();
		wordListModel = new WordListModel();
		list.setAlignmentX(JList.CENTER_ALIGNMENT);
		list.setModel(wordListModel);
		scrollPane.setViewportView(list);
		list.setBorder(null);
		list.setFont(new Font("Segoe UI", Font.PLAIN, 12));

		try {
			frmSlagalica.setIconImage(ImageIO.read(Thread.currentThread().getContextClassLoader()
					.getResourceAsStream(ICON_PATH)));

		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	public void updateTheList() {
		EventQueue.invokeLater(new Runnable() {
			public void run() {
				try {
					wordListModel.setValues(serbianDictionary
							.findAllWords(textField.getText()));
					wordListModel.refresh();
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
		});
	}

}

File: src/main/java/slagalica/gui/SlagalicaGUI.kt
@file:Suppress("SpellCheckingInspection")

package slagalica.gui

import slagalica.core.Dictionary
import slagalica.loadResource
import java.awt.*
import javax.imageio.ImageIO
import javax.swing.*

class SlagalicaGUI {
    private val serbianDictionary = Dictionary(DICTIONARY_PATH)
    private val wordListModel = WordListModel().apply {
        values = serbianDictionary.findAllWords(null)
    }

    private val textField = JTextField()
    private val frmSlagalica = initialize(textField, wordListModel)

    private fun initialize(searchField: JTextField, model: WordListModel): JFrame = JFrame().apply {
        title = "Slagalica"
        setBounds(100, 100, 578, 405)
        defaultCloseOperation = WindowConstants.EXIT_ON_CLOSE

        iconImage = ImageIO.read(loadResource(ICON_PATH))

        contentPane.apply {
            val findWords = FindWordActionListener(this@SlagalicaGUI)

            layout = GridBagLayout().apply {
                columnWidths = intArrayOf(100, 300, 100, 0)
                rowHeights = intArrayOf(0, 0, 0)
                columnWeights = doubleArrayOf(0.0, 2.0, 0.0, Double.MIN_VALUE)
                rowWeights = doubleArrayOf(0.0, 1.0, Double.MIN_VALUE)
            }

            add(JLabel("Унесите слова:"), GridBagConstraints().apply {
                insets = Insets(16, 16, 16, 16)
                anchor = GridBagConstraints.WEST

            })

            add(searchField.apply {
                columns = 10
                addActionListener(findWords)
                border = BorderFactory.createCompoundBorder(
                    BorderFactory.createLineBorder(Color.BLACK, 1, true), BorderFactory.createEmptyBorder(8, 8, 8, 8)
                )
            }, GridBagConstraints().apply {
                insets = Insets(16, 0, 16, 0)
                fill = GridBagConstraints.BOTH

            })

            add(JButton("Пронађи речи").apply {
                addActionListener(findWords)
            }, GridBagConstraints().apply {
                insets = Insets(16, 16, 16, 16)
                fill = GridBagConstraints.BOTH
            })

            add(JScrollPane().apply {
                viewportBorder = null
                border = BorderFactory.createLineBorder(Color.BLACK, 1, true)

                setViewportView(JList<String>().apply {
                    alignmentX = JList.CENTER_ALIGNMENT
                    cellRenderer = object : JLabel(), ListCellRenderer<String> {
                        override fun getListCellRendererComponent(
                            list: JList<out String>?,
                            value: String?,
                            index: Int,
                            isSelected: Boolean,
                            cellHasFocus: Boolean
                        ): Component {
                            border = BorderFactory.createEmptyBorder(4, 4, 4, 4)
                            text = value
                            return this
                        }
                    }
                    this.model = model
                    border = null
                })
            }, GridBagConstraints().apply {
                fill = GridBagConstraints.BOTH
                gridwidth = 3
                insets = Insets(0, 8, 8, 8)
                gridx = 0
                gridy = 1
            })
        }
    }

    fun updateTheList() {
        EventQueue.invokeLater {
            try {
                wordListModel.values = serbianDictionary.findAllWords(textField.text)
                wordListModel.refresh()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    companion object {
        private const val ICON_PATH = "images/app_icon.png"
        private const val DICTIONARY_PATH = "data/sr-cyrl.wl"

        @JvmStatic
        fun main(args: Array<String>) {
            Taskbar.getTaskbar().iconImage = ImageIO.read(loadResource(ICON_PATH))
            EventQueue.invokeLater {
                val window = SlagalicaGUI()
                window.frmSlagalica.isVisible = true
            }
        }
    }
}